File: enterprise/frontend/src/metabase-enterprise/audit_app/lib/cards/question_detail.js
export const viewsByTime = questionId => ({
  card: {
    name: "Views last 45 days",
    display: "bar",
    dataset_query: {
      type: "internal",
      fn: "metabase-enterprise.audit-app.pages.question-detail/cached-views-by-time",
      args: [questionId, "day"],
    },
    visualization_settings: {
      series_settings: {
        "cached-views": {
          title: "Cached",
          color: "#88BF4D",
        },
        "uncached-views": {
          title: "Uncached",
          color: "#EF8C8C",
        },
      },
      "graph.dimensions": ["date"],
      "graph.metrics": ["cached", "uncached"],
      "stackable.stack_type": "stacked",
      "graph.y_axis.title_text": "Views",
    },
  },
});

export const averageExecutionTime = questionId => ({
  card: {
    name: "Average execution time last 45 days",
    display: "line",
    dataset_query: {
      type: "internal",
      fn: "metabase-enterprise.audit-app.pages.question-detail/avg-execution-time-by-time",
      args: [questionId, "day"],
    },
  },
});

export const revisionHistory = questionId => ({
  card: {
    name: "Revision history",
    display: "table",
    dataset_query: {
      type: "internal",
      fn: "metabase-enterprise.audit-app.pages.question-detail/revision-history",
      args: [questionId],
    },
    visualization_settings: {
      "table.columns": [
        { name: "user_id", enabled: true },
        { name: "change_made", enabled: true },
        { name: "revision_id", enabled: true },
        { name: "timestamp", enabled: true, date_format: "M/D/YYYY, h:mm A" },
      ],
    },
  },
});

export const auditLog = questionId => ({
  card: {
    name: "Audit log",
    display: "table",
    dataset_query: {
      type: "internal",
      fn: "metabase-enterprise.audit-app.pages.question-detail/audit-log",
      args: [questionId],
    },
    visualization_settings: {
      "table.columns": [
        { name: "user_id", enabled: true },
        { name: "when", enabled: true },
        {
          name: "what",
          enabled: true,
          // This needs to combinatorially explore the metadata boolean space: if n grows above 2, replace mustache
          markdown_template: `
{{#json.ignore_cache}}Requested un-cached results{{/json.ignore_cache}}
{{^json.ignore_cache}}
{{#json.cached}}Viewed (cached){{/json.cached}}
{{^json.cached}}Viewed{{/json.cached}}
{{/json.ignore_cache}}`,
        },
      ],
    },
  },
});

File: enterprise/backend/src/metabase_enterprise/audit_app/pages/question_detail.clj
(ns metabase-enterprise.audit-app.pages.question-detail
  "Detail page for a single Card (Question)."
  (:require
   [metabase-enterprise.audit-app.interface :as audit.i]
   [metabase-enterprise.audit-app.pages.common :as common]
   [metabase-enterprise.audit-app.pages.common.card-and-dashboard-detail
    :as card-and-dash-detail]
   [metabase.models.card :refer [Card]]
   [metabase.util.malli :as mu]
   [metabase.util.malli.schema :as ms]))

;; Get views of a Card broken out by a time `unit`, e.g. `day` or `day-of-week`.
(mu/defmethod audit.i/internal-query ::views-by-time
  [_query-type
   card-id       :- ms/PositiveInt
   datetime-unit :- common/DateTimeUnitStr]
  (card-and-dash-detail/views-by-time "card" card-id datetime-unit))

;; Get cached views of a Card broken out by a time `unit`, e.g. `day` or `day-of-week`.
(mu/defmethod audit.i/internal-query ::cached-views-by-time
  [_query-type
   card-id       :- ms/PositiveInt
   datetime-unit :- common/DateTimeUnitStr]
  (card-and-dash-detail/cached-views-by-time card-id datetime-unit))

;; Get the revision history for a Card.
(mu/defmethod audit.i/internal-query ::revision-history
  [_query-type card-id :- ms/PositiveInt]
  (card-and-dash-detail/revision-history Card card-id))

;; Get a view log for a Card.
(mu/defmethod audit.i/internal-query ::audit-log
  [_query-type card-id :- ms/PositiveInt]
  (card-and-dash-detail/audit-log "card" card-id))

;; Average execution time broken out by period
(mu/defmethod audit.i/internal-query ::avg-execution-time-by-time
  [_query-type
   card-id       :- ms/PositiveInt
   datetime-unit :- common/DateTimeUnitStr]
  (card-and-dash-detail/avg-execution-time-by-time card-id datetime-unit))
